File: app/src/main/java/com/akilimo/mobile/views/fragments/PlantingDateFragment.java
package com.akilimo.mobile.views.fragments;


import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatDialogFragment;
import androidx.appcompat.widget.AppCompatButton;
import androidx.fragment.app.FragmentManager;

import com.akilimo.mobile.databinding.FragmentPlantingHarvestDateBinding;
import com.crashlytics.android.Crashlytics;
import com.google.android.gms.common.util.Strings;
import com.akilimo.mobile.R;
import com.akilimo.mobile.entities.ScheduledDate;
import com.akilimo.mobile.inherit.BaseStepFragment;
import com.akilimo.mobile.utils.DateHelper;
import com.akilimo.mobile.views.fragments.dialog.DateDialogPickerFragment;
import com.stepstone.stepper.VerificationError;

;

/**
 * A simple {@link androidx.fragment.app.Fragment} subclass.
 */
public class PlantingDateFragment extends BaseStepFragment {


    TextView lblSelectedPlantingDate;
    TextView lblSelectedHarvestDate;
    AppCompatButton btnPickPlantingDate;
    AppCompatButton btnPickHarvestDate;

    FragmentPlantingHarvestDateBinding binding;


    private String selectedPlantingDate;
    private String selectedHarvestDate;
    private int plantingWindow = 0;
    private int harvestWindow = 0;


    private ScheduledDate scheduledDate;

    private boolean alreadyPlanted = false;

    public PlantingDateFragment() {
        // Required empty public constructor
    }

    public static PlantingDateFragment newInstance() {
        return new PlantingDateFragment();
    }


    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        this.context = context;
    }


    @Override
    protected View loadFragmentLayout(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentPlantingHarvestDateBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        lblSelectedPlantingDate = binding.lblSelectedPlantingDate;
        lblSelectedHarvestDate = binding.lblSelectedHarvestDate;
        btnPickPlantingDate = binding.btnPickPlantingDate;
        btnPickHarvestDate = binding.btnPickHarvestDate;

        final FragmentManager fm = getActivity().getSupportFragmentManager();

        btnPickPlantingDate.setOnClickListener(v -> {
            // create the datePickerFragment
            AppCompatDialogFragment newFragment = new DateDialogPickerFragment(true);
            // set the targetFragment to receive the results, specifying the request code
            newFragment.setTargetFragment(PlantingDateFragment.this, DateDialogPickerFragment.PLANTING_REQUEST_CODE);
            // show the datePicker
            newFragment.show(fm, "PlantingDatePicker");
        });

        btnPickHarvestDate.setOnClickListener(v -> {
            // create the datePickerFragment
            AppCompatDialogFragment newFragment = new DateDialogPickerFragment(true, selectedPlantingDate);
            // set the targetFragment to receive the results, specifying the request code
            newFragment.setTargetFragment(PlantingDateFragment.this, DateDialogPickerFragment.HARVEST_REQUEST_CODE);
            // show the datePicker
            newFragment.show(fm, "HarvestDatePicker");
        });


    }

    public void refreshData() {
        try {

            scheduledDate = database.scheduleDateDao().findOne();

            if (scheduledDate != null) {
                selectedPlantingDate = scheduledDate.getPlantingDate();
                selectedHarvestDate = scheduledDate.getHarvestDate();
                lblSelectedPlantingDate.setText(selectedPlantingDate);
                lblSelectedHarvestDate.setText(selectedHarvestDate);
            }

        } catch (Exception ex) {
            Toast.makeText(context, ex.getMessage(), Toast.LENGTH_SHORT).show();
            Crashlytics.log(Log.ERROR, LOG_TAG, ex.getMessage());
            Crashlytics.logException(ex);
        }
    }


    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        // check for the results
        if (resultCode == Activity.RESULT_OK) {
            if (requestCode == DateDialogPickerFragment.PLANTING_REQUEST_CODE) {
                selectedPlantingDate = data.getStringExtra("selectedDate");
                selectedHarvestDate = null;
            } else if (requestCode == DateDialogPickerFragment.HARVEST_REQUEST_CODE) {
                selectedHarvestDate = data.getStringExtra("selectedDate");
            }
        }
        lblSelectedPlantingDate.setText(selectedPlantingDate);
        lblSelectedHarvestDate.setText(selectedHarvestDate);
    }

    private void saveEntities() {

        dataIsValid = !Strings.isEmptyOrWhitespace(selectedPlantingDate);
        if (!dataIsValid) {
            errorMessage = context.getString(R.string.lbl_planting_date_prompt);
            return;
        }
        dataIsValid = !Strings.isEmptyOrWhitespace(selectedHarvestDate);
        if (!dataIsValid) {
            errorMessage = context.getString(R.string.lbl_harvest_date_prompt);
            return;
        }
        try {


            if (scheduledDate == null) {
                scheduledDate = new ScheduledDate();
            }
            DateHelper.dateTimeFormat = "dd/MM/yyyy";
            alreadyPlanted = DateHelper.olderThanCurrent(selectedPlantingDate);

            scheduledDate.setPlantingDate(selectedPlantingDate);
            scheduledDate.setHarvestDate(selectedHarvestDate);
            scheduledDate.setAlreadyPlanted(alreadyPlanted);

            if (scheduledDate.getId() != null) {
                database.scheduleDateDao().update(scheduledDate);
            } else {
                database.scheduleDateDao().insert(scheduledDate);
            }
            scheduledDate = database.scheduleDateDao().findOne();
            dataIsValid = true;
        } catch (Exception ex) {
            dataIsValid = false;
            errorMessage = ex.getMessage();
            Crashlytics.log(Log.ERROR, LOG_TAG, ex.getMessage());
            Crashlytics.logException(ex);
        }
    }

    @Nullable
    @Override
    public VerificationError verifyStep() {
        saveEntities();
        if (!dataIsValid) {
            showCustomWarningDialog(errorMessage);
            return new VerificationError(errorMessage);
        }
        return null;
    }

    @Override
    public void onSelected() {
        refreshData();
    }

    @Override
    public void onError(@NonNull VerificationError error) {

    }
}

File: app/src/main/java/com/akilimo/mobile/views/fragments/PlantingDateFragment.kt
package com.akilimo.mobile.views.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.viewModels
import com.akilimo.mobile.databinding.FragmentPlantingHarvestDateBinding
import com.akilimo.mobile.inherit.BindBaseStepFragment
import com.akilimo.mobile.viewmodels.PlantingDateViewModel
import com.akilimo.mobile.viewmodels.factory.PlantingDateViewModelFactory
import com.akilimo.mobile.views.fragments.dialog.DateDialogPickerFragment
import com.stepstone.stepper.VerificationError


/**
 * A simple [androidx.fragment.app.Fragment] subclass.
 */
class PlantingDateFragment : BindBaseStepFragment<FragmentPlantingHarvestDateBinding>() {

    private val viewModel: PlantingDateViewModel by viewModels {
        PlantingDateViewModelFactory(requireActivity().application)
    }

    companion object {
        fun newInstance(): PlantingDateFragment {
            return PlantingDateFragment()
        }
    }

    override fun inflateBinding(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ) = FragmentPlantingHarvestDateBinding.inflate(inflater, container, false)

    override fun onBindingReady(savedInstanceState: Bundle?) {

        binding.plantingBtnPickDate.setOnClickListener { _: View? ->
            binding.harvestBtnPickDate.isEnabled = false
            DateDialogPickerFragment.newInstanceForPlanting()
                .show(parentFragmentManager, "PlantingDatePicker")
        }

        binding.harvestBtnPickDate.setOnClickListener { _: View? ->
            DateDialogPickerFragment.newInstanceForHarvest(viewModel.plantingDate.value ?: "")
                .show(parentFragmentManager, "HarvestDatePicker")
        }

        setupDateResultListeners()
        setupObservers()
    }

    private fun setupDateResultListeners() {
        parentFragmentManager.setFragmentResultListener(
            DateDialogPickerFragment.PLANTING_RESULT_KEY,
            viewLifecycleOwner
        ) { _, bundle ->
            val date = bundle.getString("selectedDate") ?: ""
            viewModel.setPlantingDate(date)
        }

        parentFragmentManager.setFragmentResultListener(
            DateDialogPickerFragment.HARVEST_RESULT_KEY,
            viewLifecycleOwner
        ) { _, bundle ->
            val date = bundle.getString("selectedDate") ?: ""
            viewModel.setHarvestDate(date)
        }
    }

    override fun setupObservers() {
        viewModel.plantingDate.observe(viewLifecycleOwner) { plantingDate ->
            binding.plantingDateLabel.text = plantingDate
            binding.harvestBtnPickDate.isEnabled = plantingDate.isNotEmpty()
        }

        viewModel.harvestDate.observe(viewLifecycleOwner) { harvestDate ->
            binding.harvestDateLabel.text = harvestDate
        }

        viewModel.errorMessage.observe(viewLifecycleOwner) {
            it?.let { msg -> Toast.makeText(requireContext(), msg, Toast.LENGTH_SHORT).show() }
        }
    }

    override fun onSelected() {
        viewModel.loadInitialDates()
    }
    override fun verifyStep(): VerificationError? {
        viewModel.saveSchedule()

        val error = viewModel.errorMessage.value
        return if (error.isNullOrBlank()) {
            null
        } else {
            showCustomWarningDialog(error)
            VerificationError(error)
        }
    }
}
